### PR TITLE
Expose search sort parameters in HAL-FORMS properties

### PR DESCRIPTION
**Pull request by vierbergenlars**
To be able to add the sort parameter to HAL-FORMS properties (and potentially other properties in the future), first the HAL-FORMS template generation is expanded to allow multiple contributors to add fields.

Then, this infrastructure is used to add the `sort` parameter with all sortable fields as HAL-FORMS options.
A standard HAL-FORMS client will display the prompt in a (multi)select. The prompt describes which field and the order in which it is sorted.

A custom client can recognize the `sort` field (by name) and handle the full options objects (using `property` and `direction` to determine which fields map to which options)

The profile response for the search form looks like this:

```json 
{
 "_templates": {
    "search": {
      "method": "GET",
      "properties": [
        {
          "name": "document.length",
          "type": "number"
        },
        {
          "name": "document.mimetype",
          "prompt": "Document mimetype",
          "type": "text"
        },
        {
          "name": "pay_before",
          "prompt": "Pay before",
          "type": "datetime"
        },
        {
          "name": "pay_before~before",
          "type": "datetime"
        },
        {
          "name": "pay_before~after",
          "type": "datetime"
        },
        {
          "name": "total_amount~lte",
          "type": "number"
        },
        {
          "name": "total_amount~gte",
          "type": "number"
        },
        {
          "name": "supplier.name",
          "prompt": "Supplier: Name",
          "type": "text",
          "options": {
            "maxItems": 1,
            "inline": [
              "Xenit",
              "AB Inbev"
            ]
          }
        },
        {
          "name": "supplier.bank_account",
          "prompt": "Supplier: Bank account",
          "type": "text"
        },
        {
          "name": "supplier.bank_account~ci",
          "type": "text"
        },
        {
          "name": "sort",
          "type": "text",
          "options": {
            "promptField": "prompt",
            "valueField": "value",
            "minItems": 0,
            "inline": [
              {
                "property": "document.length",
                "direction": "asc",
                "prompt": "document.length 0🡲9",
                "value": "document.length,asc"
              },
              {
                "property": "document.length",
                "direction": "desc",
                "prompt": "document.length 9🡲0",
                "value": "document.length,desc"
              },
              {
                "property": "document.mimetype",
                "direction": "asc",
                "prompt": "Document mimetype A🡲Z",
                "value": "document.mimetype,asc"
              },
              {
                "property": "document.mimetype",
                "direction": "desc",
                "prompt": "Document mimetype Z🡲A",
                "value": "document.mimetype,desc"
              },
              {
                "property": "pay_before",
                "direction": "asc",
                "prompt": "Pay before oldest first",
                "value": "pay_before,asc"
              },
              {
                "property": "pay_before",
                "direction": "desc",
                "prompt": "Pay before newest first",
                "value": "pay_before,desc"
              }
            ]
          }
        }
      ],
      "target": "http://localhost:9000/invoices"
    }
}
```

### CODE REVIEW

File: contentgrid-spring-data-rest/src/main/java/com/contentgrid/spring/data/querydsl/mapping/CollectionFilterHalFormsPayloadMetadataContributor.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    @Override
+    public Stream<PropertyMetadata> contributeToUpdateForm(Class<?> domainType) {
+        return Stream.empty();
+    }
+
+    @Override
+    public Stream<PropertyMetadata> contributeToSearchForm(Class<?> domainType) {
+        return collectionFiltersMapping.forDomainType(domainType)
+                .filters()
+                .filter(CollectionFilter::isDocumented)
```
Comment (NielsCW) [inline contentgrid-spring-data-rest/src/main/java/com/contentgrid/spring/data/querydsl/mapping/CollectionFilterHalFormsPayloadMetadataContributor.java:31]: It is possible to use `CollectionFilters#Documented()` here

File: contentgrid-spring-data-rest/src/main/java/com/contentgrid/spring/data/querydsl/sort/CollectionFilterSortHalFormsPayloadMetadataContributor.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+                    return HalFormsOptions.inline(sortPropertyMetadata.getSortOptions())
+                            .withMinItems(0L)
+                            .withValueField("value")
+                            .withPromptField("prompt");
+                }
+
+                // Returning null here will just not add any options to this field.
+                // That is the expected fallback when the sortParam does turn out not to actually be SortPropertyMetadata,
+                // since that means the sort parameter was actually an existing field name on the domain type
+                return null;
```
Comment (NielsCW) [inline contentgrid-spring-data-rest/src/main/java/com/contentgrid/spring/data/querydsl/sort/CollectionFilterSortHalFormsPayloadMetadataContributor.java:85]: Is this covered in tests?

File: contentgrid-spring-data-rest/src/test/java/com/contentgrid/spring/data/rest/messages/HalLinkTitlesAndFormPromptsTest.java [hunk 3]
```diff
@@ -83,7 +83,64 @@ void promptOnCreateFormPropertiesInHalForms() throws Exception {
...
+                                        { name: "invoices.number", type: "text" }, { name: "invoices.paid", type: "checkbox" }, { name: "invoices.orders.id" }, { name: "invoices.content.length", type: "number" },
+                                        {
+                                            name: "sort",
+                                            options: {
+                                                promptField: "prompt",
+                                                valueField: "value",
+                                                inline: [
+                                                    {
+                                                        value: "vat,asc",
+                                                        prompt: "VAT number A🡲Z" # Note that the field value is replaced with the proper translation
```
Comment (NielsCW) [inline contentgrid-spring-data-rest/src/test/java/com/contentgrid/spring/data/rest/messages/HalLinkTitlesAndFormPromptsTest.java:95]: Tests are not updated by commit `Replace arrow with one that is more commonly present in fonts`

File: contentgrid-spring-data-rest/src/main/java/com/contentgrid/spring/data/querydsl/sort/CollectionFilterSortHalFormsPayloadMetadataContributor.java (hunk 2 above)
Comment (vierbergenlars) [reply contentgrid-spring-data-rest/src/main/java/com/contentgrid/spring/data/querydsl/sort/CollectionFilterSortHalFormsPayloadMetadataContributor.java:85]: It is covered now in a new test that has an entity with a `sort` property (which is not searchable)

**Review by NielsCW — APPROVED** [review]

**Pull request merged**